Bs10 filter by year

stumbo opened the pull request:
A revision to my initial work.  This version still needs testing, but seems to be closer to the mark.  

The date field now support a single date, `1980`; a comparison, ` < 1990`; or a range, `1990 between 2000`.  I need to replace the word between - it sounded good in my initial thoughts but, in reality, it seems kind of clunky.  

To do - testing of edge cases.

This version is live on the [staging site](https://stumbo.github.io/InterlispDraft.github.io/history/bibliography/)

stumbo committed:
Initial implementation of adding filter by year

stumbo committed:
Update dates to simple expressions

stumbo committed:
Add new version of date selection

This version replaces using tags with a text window that allows
entering single dates, a relationship (<,>, etc.) and a date or
the phrase `date between date` allowing a selection of dates.

nbriggs commented:
How about `1990 to 2000`, or `1990-2000` for the range?

hjellinek commented:
> How about `1990 to 2000`, or `1990-2000` for the range?

Or allow both?

stumbo committed:
Change 'between' to 'to'.  Fix comparison to use start of year.

stumbo committed:
Updates to fix comparison operators

stumbo committed:
Remove unneeded JSON creation.  More comparison operator cleanup.

stumbo committed:
Add tooltip, reformatting leverage bootstrap, less custom css

stumbo commented:
Significant revisions since I put up the draft.  I went with just `to` for specifying a range.  With `-` also being the separator for dates, it seemed like too many dashes.

Good dates to experiment with 1959, 1959-09, 1959-09-01, 1987, 1987-06, and 1987-06-01.

1987-06-01 points out an idiosyncrasy with attempting to compare various date formats.  There are a bunch of items published in June 1987, most with just the month and year, but one with the full date.  searching on 1987-06-01 returns all the papers published with just June 1987 as the date and the paper published on the first.  If you switch to 1987-06-02, all the June 1987 papers are filtered out.  To do comparisons I expand the date to a full date and assume the first month or day of the year for missing parts.

I'm tempted to leave it alone and put it on the list of future improvements.

I also added a tooltip to help provide information on legal values for the date field.  Let me know what you think of it.

The current code is running on the [staging site](https://stumbo.github.io/InterlispDraft.github.io/history/bibliography/).

hjellinek commented:
I like the capability, but the interaction style was a little surprising.  The filter runs as you type, which is fast and cool, but when I finish typing the date expression and hit the return key, as is my habit, the date field clears, the filter is removed, and I'm looking at all the documents instead of those in the date range I just typed.

pamoroso commented:
The return key behavior surprised me too.

masinter commented:
the handling of partial dates by extending the date specification to a specific date (January 1 or December 31) doesn't really work, because the query date start vs. query date end are different --   if you say "1998" you need to include anything published between 1998-01-01 and 1998-12-31.    There are multiple dates: date of publication vs. date of subject matter (a paper about Interlisp-10 written in 2025 and published in 2025 but the topic is the system between 1970 and 1984, for example).

stumbo committed:
Stop date field for erasing on CR

stumbo commented:
> I like the capability, but the interaction style was a little surprising. The filter runs as you type, which is fast and cool, but when I finish typing the date expression and hit the return key, as is my habit, the date field clears, the filter is removed, and I'm looking at all the documents instead of those in the date range I just typed.

Good catch - my usage has been just typing in the field and seeing the results.  I never hit return. 

I've fixed the behavior.  A return no longer clears the field, it just leaves the type-in point where it is.  Does that behavior better match your expectations?

stumbo commented:
> the handling of partial dates by extending the date specification to a specific date (January 1 or December 31) doesn't really work, because the query date start vs. query date end are different -- if you say "1998" you need to include anything published between 1998-01-01 and 1998-12-31. There are multiple dates: date of publication vs. date of subject matter (a paper about Interlisp-10 written in 2025 and published in 2025 but the topic is the system between 1970 and 1984, for example).

I think you're poking at two separate issues.  Right now I'm only addressing publication date.  Addressing subject matter Interlisp-10 or some other version will need to be covered by good metadata on the paper.  That can come either from tags entered in Zotero or using some other tool to summarize and extract key phrases from documents.  Definitely items we should investigate.

Back to publications dates - maybe the best solution is to focus only on years.  Is adding in the greater specificity of months and days buying any real value?  If we just focus on years, it will simplify the comparison and the little details of partial publication dates can be ignored, it was either published in a year or not.

pamoroso commented:
The ENTER key now works well.

stumbo committed:
Overhaul to focus on just years.

Input is limited to just years and comparison operators.
Changed so processing of input occurs on CR.  Include
validation of input.  Revised Tooltip to hide once user
starts entering input.

stumbo commented:
Revised Date filtering to just focus on years - renamed field to *Filter by Year*, changed processing to only accept years, perform validation and filter operation on CR.  Invalid input will turn the box red on CR.

Current version has been deployed to the staging site.

pamoroso commented:
Hoovering over the date field I get this tooltip with HTML code:

<img width="956" height="183" alt="date-field" src="https://github.com/user-attachments/assets/0f1c861b-5f81-43cc-bee5-cfd6ebb65abf" />

This happens with Firefox but not Chrome.

pamoroso commented:
I forgot: Firefox on Linux.

hjellinek reviewed (approved):
I can't comment on the code, but the behavior and appearance are both what I'd expect.  No surprises!  Works well and looks good, in other words.

stumbo commented:
> Hoovering over the date field I get this tooltip with HTML code:
> <img alt="date-field" width="956" height="183" src="https://private-user-images.githubusercontent.com/10342479/519208720-0f1c861b-5f81-43cc-bee5-cfd6ebb65abf.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQyMTk5ODEsIm5iZiI6MTc2NDIxOTY4MSwicGF0aCI6Ii8xMDM0MjQ3OS81MTkyMDg3MjAtMGYxYzg2MWItNWY4MS00M2NjLWJlZTUtY2ZkNmViYjY1YWJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTI3VDA1MDEyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0ZGU2ODRiYjQ1YTIzZWRhNjQ1NjcyZWFhOGExNzE4NGQ0M2VkZTVmMzM5NmQ2ZTY4YmRkOWExZTgwNDIwN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KodQN2N9SPDbaTVfgkE4hIfS-F9yNQG_v5e93gaWFzw">
> 
> This happens with Firefox but not Chrome.

Can you try this again?  I haven't been able to duplicate the issue on my Windows, Mac or Ubuntu machines.  Nor does it occur in a private window.

pamoroso commented:
I can no longer reproduce the issue.

stumbo commented:
> I can no longer reproduce the issue.

Ok - thanks.  I suspect this issue you had was due to the browser cache not being refreshed.  The problem (at least when I saw it) was caused by how I was generating the tooltip.  I had revamped that to fix it in the latest version, you likely had a stale version.

I'm going to go ahead and merge - if we see it in the future I'll investigate further.
